types/src/zkid.rs

```diff
@@ -0,0 +1,429 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    jwks::rsa::RSA_JWK,
+    on_chain_config::CurrentTimeMicroseconds,
+    transaction::{
+        authenticator::{
+            AnyPublicKey, AnySignature, EphemeralPublicKey, EphemeralSignature, MAX_NUM_OF_SIGS,
+        },
+        SignedTransaction,
+    },
+};
+use anyhow::{anyhow, ensure, Context, Ok, Result};
+use aptos_crypto::{poseidon_bn254, CryptoMaterialError, ValidCryptoMaterial};
+use ark_bn254;
+use ark_serialize::CanonicalSerialize;
+use base64::{URL_SAFE, URL_SAFE_NO_PAD};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use serde_with::skip_serializing_none;
+use std::{
+    collections::BTreeMap,
+    str,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+pub const PEPPER_NUM_BYTES: usize = poseidon_bn254::BYTES_PACKED_PER_SCALAR;
+pub const EPK_BLINDER_NUM_BYTES: usize = poseidon_bn254::BYTES_PACKED_PER_SCALAR;
+pub const NONCE_NUM_BYTES: usize = 32;
+pub const IDC_NUM_BYTES: usize = 32;
+
+// TODO(ZkIdGroth16Zkp): add some static asserts here that these don't exceed the MAX poseidon input sizes
+// TODO(ZkIdGroth16Zkp): determine what our circuit will accept
+
+pub const MAX_EPK_BYTES: usize = 93; // Supports public key lengths of up to 92 bytes.
+pub const MAX_ISS_BYTES: usize = 248;
+pub const MAX_AUD_VAL_BYTES: usize = 248;
+pub const MAX_UID_KEY_BYTES: usize = 248;
+pub const MAX_UID_VAL_BYTES: usize = 248;
+pub const MAX_JWT_HEADER_BYTES: usize = 248;
+
+pub const MAX_ZK_PUBLIC_KEY_BYTES: usize = MAX_ISS_BYTES + MAX_EPK_BYTES;
+
+// TODO(ZkIdGroth16Zkp): determine max length of zkSNARK + OIDC overhead + ephemeral pubkey and signature
+pub const MAX_ZK_SIGNATURE_BYTES: usize = 1024;
+
+// TODO(ZkIdGroth16Zkp): each zkID Groth16 proof will take ~2 ms to verify, or so. We cannot verify too many due to DoS.
+//  Figure out what this should be.
+pub const MAX_ZK_ID_AUTHENTICATORS_ALLOWED: usize = 10;
+
+// How far in the future from the JWT issued at time the EPK expiry can be set.
+pub const MAX_EXPIRY_HORIZON: u64 = 1728000000; // 2000 days TODO(zkid): finalize this value
```

Comment:
I haven't crunched the numbers, but if these are seconds, google says it's 20,000 days, not 2,000.